### DISCUSSION TITLE
CPU utilization increased after using AWS mysql JDBC driver

### DISCUSSION THREAD

**Discussion by jkrmelj** [awslabs/aws-mysql-jdbc#186] [Q&A]
Hi,

we are using AWS Mysql compatible Aurora and recently we switched from using mariaDb driver to AWS mysql compatible driver. The transition was smooth in the sense of functionality.

But after the deploy of the application with the new AWS driver enabled we noticed an increase of CPU usage within the application. I'm attaching the graph of the CPU usage, after the deploy of new version.

<img width="1539" alt="Screenshot 2022-04-06 at 18 31 36" src="https://user-images.githubusercontent.com/1469484/162024204-49130ec8-a9a6-4353-869e-0cf22733b1d8.png">

While searching through the documentation of the AWS driver, I couldn't find any explanation for this increase. Can anyone explain where this increase is coming from and are there any possible mitigations to decrease the CPU load?

Thank you!

**Comment by sergiyvamz** [comment]
Hello @jkrmelj,

Thank you for giving the AWS JDBC Driver for MySQL a try!

The latest release of the driver (1.0.0) has introduced a new Plugin Manager feature. The main benefit of the Plugin Manager is the ability to load, or not to load, standard and user custom plugins, thus extending driver functionality. By default, the driver loads the Failover plugin and Enhanced Failure Monitoring plugin.

For reference, please refer to this section in our README: https://github.com/awslabs/aws-mysql-jdbc#:~:text=Failover%20and%20Enhanced%20Failure%20Monitoring%20are%20loaded%20by%20default.

The Failure Monitoring plugin helps to detect network outages and unexpected DB cluster failover events more efficiently. However, it comes with a small price. Enhanced Failure Monitoring plugin runs a monitoring thread and proactively checks node health status. Despite the fact that monitoring thread is very light, it adds some load to the CPU. I assume that is the reason you see CPU usage increase after migration to the driver.

You may consider to not load Enhanced Failure Monitoring plugin and verify that it's the reason of the increased CPU utilization. Check out details about `connectionPluginFactories` parameter at: https://github.com/awslabs/aws-mysql-jdbc#connection-plugin-manager-parameters.

If the above does not provide further explanation, another reason why the difference might be seen is that the AWS JDBC Driver for MySQL is based off the community MySQL JDBC driver, so there may be inherent differences within the codebase that might differ with CPU utilization when compared to MariaDB.

**Comment by jkrmelj** [comment]
Hi,

thank you for your feedback. According to the suggestions you provided I tried to disable the plugins:
1. I disabled the Failover plugin only by setting enableClusterAwareFailover=false. The CPU usage on an application side did not change in any manner.
2. Then I disabled loading of all plugins by setting useConnectionPlugins=false. The CPU usage on an application side did fell by few percents. So this confirm your theory Failure Monitoring plugin is a lightweight one and does not affect client much.

If I disable AWS mysql driver and switch back using mariaDb one, the CPU usage fells down significantly. So it seems root cause lies in the underlying DB driver.

The application itself is write intensive, producing around 6k updates per second. We are currently writing into four Aurora clusters in parallel. Therefore application is using four (Hikari) connection pools at a time, one for each cluster. Could this be the reason? Handling same driver within multiple connection pools?

Performance Insights for the Aurora clusters do not show anything unusual. CPU usage on DB side is stable and does not differ between usage of AWS and mariaDb driver. Also no additional queries are seen on the DB side.
